---
title: "Network Analysis by TB Secreted Factor"
output: html_document
started: "2023-06-15"
updated: "2023-06-15"
---


```{r, Dependencies}
library (data.table)
#library (ComplexHeatmap)
#library (ggplot2)
source("..\\Helper_R_Scripts\\shortestPathFunctions.R")

source("..\\Helper_R_Scripts\\ManageScriptData.R")
WriteSessionInfo()
WriteInstalledPackages()
```


```{r, Import Data}
log2fct <- 1    # thresholds, conventional
fdrt <- 0.05

esxProteins <- c("ESAT6", "Espb", "ESPR",  "Esxq","Pe25", "Pe35", "Ppe11")

# Import Mehdi's subnet from netprop
npout <- fread("..\\Input_Data\\MehdiNP_outputs_nonorm.txt") # x0 is input heat, xss_t is propagated heat
npin <- fread("..\\Input_Data\\MehdiNP_inputs.txt")
npout <- npout[pval < fdrt]


# Import STRING PPI network and aliases
#stringNet <- fread("https://stringdb-static.org/download/protein.links.detailed.v11.5/9606.protein.links.detailed.v11.5.txt.gz")
stringNet <- fread("C:\\Users\\isido\\OneDrive\\Documents\\Krogan Lab Code\\Universal_Datasets\\STRING_files\\Human_9606.protein.links.detailed.v11.5.txt")
stringSub <- stringNet[experimental >= 400]
#aliases <- fread("https://stringdb-static.org/download/protein.aliases.v11.5/9606.protein.aliases.v11.5.txt.gz")
aliases <- fread("C:\\Users\\isido\\OneDrive\\Documents\\Krogan Lab Code\\Universal_Datasets\\STRING_files\\Human_9606.protein.aliases.v11.5.txt")
setnames(aliases, "#string_protein_id", "stringID")


# Import CRISPR KD Data 
load("..\\..\\TBRU_integration_data/data_clean/Cox_model/Cox_NG_screen_model.RData")
load("..\\..\\TBRU_integration_data/data_clean/Cox_model/Cox_AWR_screen_model.RData")
awrKDs <- data.table(awr_cox)[abs(log2FC) >= log2fct & FDR <= fdrt, .(symbol, log2FC), by = contrast]
ngKDs <- data.table(ng_cox)[abs(log2FC) >= log2fct & FDR <= fdrt, .(symbol, log2FC), by = contrast]
# need to convert to human homologs !!!!
library(orthogene)
mappedKDs <- data.table(orthogene::convert_orthologs(rbind(awrKDs, ngKDs), 
                                        gene_input = "symbol",
                                        gene_output = "columns",
                                        input_species = "mouse",
                                        output_species = "human",
                                        non121_strategy = "keep_popular",
                                        verbose = TRUE) )


# Import TB secreted Factor PPI
tbPPI <- fread("..\\Input_Data\\TBsecretedAPMS_NIHMS1503192-supplement-5.csv")
setnames(tbPPI, c("Bacterial Bait  (Locus ID)", "Human Protein Uniprot", "Human Gene name"), c("Bait", "Prey", "GeneNames")) 
tbPPI <- tbPPI[ MIST >= 0.7, ]
tbPPI <- tbPPI[, .(Prey, GeneNames), by = Bait]
tbPPI[, GeneName := tstrsplit(tbPPI$GeneNames, " ", keep = 1)]
```


```{r, Prepare gene lists for Network}
# Map STRING IDs to other datasets (aliases file makes it very easy to map from anything else to STRING)
npout[aliases[alias %in% npout$nodes,], string := i.stringID, on = .(nodes = alias)]
mappedKDs[aliases[alias %in% mappedKDs$ortholog_gene,], string := i.stringID, on = .(ortholog_gene = alias)]
tbPPI[aliases[alias %in% tbPPI$Prey,], string := i.stringID, on = .(Prey = alias)]

# extract PPI 
tbSecreted <- unique(tbPPI$Bait)
tbInteractor <- unique(tbPPI$string)

# map STRING IDs to symbols 
strings2symbols <- rbind(npout[, .(string, symbol = nodes)], mappedKDs[, .(string, symbol = ortholog_gene)], tbPPI[, .(string, symbol = GeneName)])
add2Net <- c(npout$string, mappedKDs$string, tbPPI$string)

net <- stringSub[ protein1 %in% add2Net & protein2 %in% add2Net]
```


```{r, Igraphification}
library(igraph)

graph <- igraph::graph_from_edgelist(as.matrix(stringSub[, .(protein1, protein2)]))
distances <- igraph::distances(graph)
```


```{r, Simple PPI & CRISPR Neighbors}
# Genes in PPI and Crispr
genesInBoth <- mappedKDs[string %in% tbPPI$string,] [tbPPI, Bait := i.Bait, on = .(string = string) ]
fwrite(genesInBoth, "genes_in_PPIandCRISPR.csv")

# PPI and Crispr interactor directly with each other
directInteractors <- net[protein1 %in% tbPPI$string & protein2 %in% mappedKDs$string, .(protein1,protein2)]
directInteractors[tbPPI, c("Bait", "Prey", "GeneName") := .(i.Bait, i.Prey, i.GeneName), on = .(protein1 = string)]
directInteractors[mappedKDs, c("crisprKD", "crisprContrast") := .(i.ortholog_gene, i.contrast), on = .(protein2 = string)]
directInteractors[, .N, by = Bait]
fwrite(directInteractors, "direct_interactors_PPIandCRISPR.csv")

# One-Hop Interactors and Intermediate nodes

twoHopTable <- do.call(rbind, lapply( tbPPI$string, FUN = function(rn){
  distancesSubbed <- distances[rownames(distances) == rn, rownames(distances) %in% mappedKDs$string]
  twoHops <- names(distancesSubbed[distancesSubbed == 2])
  if( is.character(twoHops) ){
    return(data.table(PPI = rn, KD = twoHops))
  }
}))
twoHopTable <- twoHopTable[!is.na(KD)]

twoHopPaths <- do.call( rbind, lapply(1:nrow(twoHopTable), FUN = function(i){
  
  tempPath <- igraph::shortest_paths(graph, from = twoHopTable[i, PPI], to = twoHopTable[i, KD])
  names <- names(tempPath$vpath[[1]])
  return( data.table( PPI = names[1], Interactor = names[2], Crispr = names[3]) )
}))
  
twoHopPaths[tbPPI, PPI_symbol := i.GeneName, on = .(PPI = string)] [npout, Interactor_symbol := i.nodes, on = .(Interactor = string)] [mappedKDs, Crispr_symbol := i.symbol, on = .(Crispr = string)]

twoHopPaths[tbPPI, PPI_bait := i.Bait, on = .(PPI = string)] [npout, Interactor_dataset := i.dataset, on = .(Interactor = string)] [mappedKDs, Crispr_contrast := tstrsplit(i.contrast, "-", keep = 1), on = .(Crispr = string)]

fwrite(twoHopPaths, "twoHopGenes_in_PPIandCRISPR.csv")

## GO BACK AND ADD FDR PVALUE BASED ON LINKER GENE DEGREE IN INITIAL GRAPH
```


```{r, all KDs to nearest PPI SP Subnet}

# The goal is to track start and end nodes for each path and append netProp dataset for each intermediate node

# Standard measures for each subnet:
# 1) Extract high betweenness, high npaths, & freq endpoint genes 2) Enrich gene groups for each origin in network 3) Thresh enrichment with high centrality 4) Modularize and enrich 5) Generate network and simple chart views to show centrality scores, 'inpaths' clinical data origin, and freq endpoint genes

vsAllPPI.paths <- find_ShortestPaths(mappedKDs[string %in% rownames(distances), string], tbPPI[string %in% rownames(distances), string], graph, distances, trackStartEnds = TRUE)

startProportions <- vsAllPPI.paths$startEnds[, .N, by = .(start, inpath)] [!is.na(inpath) ]
#starts have N more than 1..???
endProportions <- vsAllPPI.paths$startEnds[, .N, by = .(end, inpath)] [!is.na(inpath) ]

vsAllPPI.info <- gen_NodeInfo(vsAllPPI.paths)
vsAllPPI.sim <- gen_RandomShortestPaths(nrow(mappedKDs), length(tbPPI[string %in% rownames(distances), string]), 1000, distances, graph)
vsAllPPI.info <- normalize_ShortestPaths(vsAllPPI.sim, vsAllPPI.info)

vsAllPPI.info [, source := "inpath"] [string %in% mappedKDs$string, source := "crispr"] [string %in% tbPPI$string, source := "ppi"]  [string %in% tbPPI$string & string %in% mappedKDs$string, source := "both"] 
vsAllPPI.info [aliases[source == "Ensembl_EntrezGene"], symbol := i.alias, on = .(string = stringID)] 
#[strings2symbols, symbol := i.symbol, on = .(string)] [, symbol := tstrsplit(symbol, split = ";| " , keep = 1)]

#vsAllPPI.info [npout, netPropOrigin := i.dataset, on = .(string = string)]


```


```{r, Export ranked lists}
# Ordered by Betweenness and Number of paths, mapped to baits, mapped to og netpropped datasets
roughBaitMap <- copy(vsAllPPI.paths$startEnds)
roughBaitMap[is.na(inpath), inpath := start] 
roughBaitMap <- roughBaitMap[, .(ppi = end, otherNode = inpath)]
roughBaitMap[ tbPPI, bait := i.Bait, on = .(ppi = string)]
roughBaitMap[, baitBias := .N, by = .(otherNode, bait)]
setorder(roughBaitMap, baitBias)

esxMap <- roughBaitMap[bait %in% esxProteins]


baitInfo <- copy(vsAllPPI.info)
baitInfo[tbPPI, BaitInteractor := i.Bait, on = .(string = string)]
baitInfo[ roughBaitMap, roughBaitPath := i.bait, on = .(string = ppi)] [ roughBaitMap, roughBaitPath := i.bait, on = .(string = otherNode)]
baitInfo[esxMap, roughBaitPath := i.bait, on = .(string = ppi)][esxMap, roughBaitPath := i.bait, on = .(string = otherNode)]

baitInfoCut <- copy(baitInfo)
baitInfoCut <- baitInfoCut[, .(string, roughBaitPath)]
               
               

orderByBetw <- copy(vsAllPPI.info)
setorder(orderByBetw, -normBetwn) 
#fwrite(orderByBetw, "subnet_InfoTable_SortedBy_Betweenness.csv")


orderByNpaths <- copy(vsAllPPI.info)
setorder(orderByNpaths, -normPaths) 
#fwrite(orderByNpaths, "subnet_InfoTable_SortedBy_NumberOfPaths.csv")


```


```{r, Gene Set Enrichment}
# compare all vs ppi vs crispr vs linker genes
# do above but threshold for normalized betweenness and normalized npaths
# compare enrichment subdivided by esx-1 interacting pathways

baitInfo[,topBetw := F][ string %in% orderByBetw[1:100, string], topBetw := T]
baitInfo[,topNpath := F][ string %in% orderByNpaths[1:100, string], topNpath := T]


gg <- baitInfo[, .(string, group = source)]
gg <- rbind(gg, baitInfo[source == "both", .(string, group = "crispr")],  baitInfo[source == "both", .(string, group = "ppi")])
gg <- gg[!grepl("both",gg$group), ]
gg <- rbind(gg, data.table( string = unique(gg$string), group = "all") )

gg2 <- copy(gg)
gg2 <- rbind(gg2, gg2[string %in% orderByBetw[1:100, string], .(string, group = paste0(group, "_highPaths"))])


gg3 <- copy(gg)
gg3 <- rbind(gg3, gg3[string %in% orderByNpaths[1:100, string], .(string, group = paste0(group, "_highBetw"))])


gg4 <- copy(gg)
gg4 <- rbind(gg4, gg4[string %in% baitInfo[roughBaitPath %in% esxProteins, string], .(string, group = paste0(group, "_esx1Path"))])


gg5 <- do.call(rbind, list(gg2, gg3, gg4))
gg5 <- gg5[grepl("all_", gg5$group)]
gg5 <- rbind(gg5, gg[grepl("all", gg$group)])



vsSource.gg <- prepare_subnetworkForEnrichment(gg, "string", "group", reassign = F)
vsSource.en <- calculate_subnetworkEnrichment(vsSource.gg)


vsPaths.gg <- prepare_subnetworkForEnrichment(gg2, "string", "group", reassign = F)
vsPaths.en <- calculate_subnetworkEnrichment(vsPaths.gg)


vsBetw.gg <- prepare_subnetworkForEnrichment(gg3, "string", "group", reassign = F)
vsBetw.en <- calculate_subnetworkEnrichment(vsBetw.gg)


vsesx1.gg <- prepare_subnetworkForEnrichment(gg4, "string", "group", reassign = F)
vsesx1.en <- calculate_subnetworkEnrichment(vsesx1.gg)

vsComparison.gg <- prepare_subnetworkForEnrichment(gg5, "string", "group", reassign = F)
vsComparison.en <- calculate_subnetworkEnrichment(vsComparison.gg)


# PLOT ALL Enrichments
plot_subnetworkEnrichments(vsComparison.en, topn = 4, save = F,  name = "vs All PPI", dims = c(10,7))
                           , 
                           #negCols = unique(vsContrast.gg[grepl("low|neg", vsContrast.gg$Group), Group]), 
                           column_split = tstrsplit(sort(unique(vsSource.gg$group)), "_|O", keep = 1) [[1]] )

plot_subnetworkEnrichments(vsPaths.en, topn = 4, save = T,  name = "threshholded to number of paths", dims = c(12,8))

plot_subnetworkEnrichments(vsBetw.en, topn = 4, save = T,  name = "threshholded to betweenness", dims = c(12,8))




```


```{r, modularize by network community}



```


```{r, Send to Cytoscape!}

vsAllPPI.paths$edgemat <- matrix(vsAllPPI.paths$edges, ncol = 2, byrow = T)
vsAllPPI.graph <- simplify(igraph::graph_from_edgelist(vsAllPPI.paths$edgemat) )

library(RCy3)
createNetworkFromIgraph(vsAllPPI.graph)
loadTableData(vsAllPPI.info, data.key.column = "string", table = "node")

loadTableData(baitInfoCut, data.key.column = "string", table = "node")

```

```{r, Manipulate Cytoscape Edges}

edgetable <- fread("..\\Selected Results\\cytoscapeEdges1.csv") 

for (x in esxProteins){
  xnodes <- baitInfoCut[roughBaitPath == x, string]
  edgetable[source %in% xnodes & target %in% xnodes, roughBaitEdge := x]
}
edgetableCut <- edgetable[, .(name, roughBaitEdge)]
loadTableData(edgetableCut, data.key.column = "name", table = "edge")

```









